rust/pattern/constraint/thing_constraint.rs

```diff
@@ -0,0 +1,358 @@
+            _ => todo!(),
+        }
+    }
+}
+
+impl Display for Predicate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Predicate::*;
+        write!(
+            f,
+            "{}",
+            match self {
+                Eq => "=",
+                Neq => "!=",
+                _ => todo!(),
+            }
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Value {
+    Long(i64),
+    Double(f64),
+    Boolean(bool),
+    String(String),
+    DateTime(NaiveDateTime),
+    Variable(Box<ThingVariable>),
+}
+impl Eq for Value {} // can't derive, because f32 does not implement Eq
+
+impl From<i64> for Value {
+    fn from(int: i64) -> Value {
+        Value::Long(int)
+    }
+}
+impl From<&str> for Value {
+    fn from(string: &str) -> Value {
+        Value::String(String::from(string))
+    }
+}
+impl From<String> for Value {
+    fn from(string: String) -> Value {
+        Value::String(string)
+    }
+}
+
+impl TryFrom<NaiveDateTime> for Value {
+    type Error = ErrorMessage;
+
+    fn try_from(date_time: NaiveDateTime) -> Result<Value, ErrorMessage> {
+        if date_time.nanosecond() % 1000000 > 0 {
+            return Err(
+                INVALID_CONSTRAINT_DATETIME_PRECISION.format(&[date_time.to_string().as_str()])
+            );
+        }
+        Ok(Value::DateTime(date_time))
+    }
+}
+
+impl From<UnboundVariable> for Value {
+    fn from(variable: UnboundVariable) -> Value {
+        Value::Variable(Box::new(variable.into_thing()))
+    }
+}
+impl From<ThingVariable> for Value {
+    fn from(variable: ThingVariable) -> Value {
+        Value::Variable(Box::new(variable))
+    }
+}
+impl From<Variable> for Value {
+    fn from(variable: Variable) -> Value {
+        Value::Variable(Box::new(variable.into_thing()))
+    }
+}
+
+impl Display for Value {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Value::*;
+        match self {
+            DateTime(date_time) => write!(f, "{}", {
+                if date_time.time().nanosecond() > 0 {
+                    date_time.format("%Y-%m-%dT%H:%M:%S.%3f")
+                } else if date_time.time().second() > 0 {
+                    date_time.format("%Y-%m-%dT%H:%M:%S")
+                } else {
+                    date_time.format("%Y-%m-%dT%H:%M")
+                }
+            }),
+            String(string) => write!(f, "\"{}\"", string),
+            Variable(var) => write!(f, "{}", var.reference),
+            _ => panic!(""),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RelationConstraint {
+    role_players: Vec<RolePlayerConstraint>,
+    scope: String,
+}
+
+impl RelationConstraint {
+    pub fn new(role_players: Vec<RolePlayerConstraint>) -> Self {
+        RelationConstraint { role_players, scope: String::from("relation") }
+    }
+
+    pub fn add(&mut self, role_player: RolePlayerConstraint) {
+        self.role_players.push(role_player);
+    }
+}
```

Comment:
Chain `rel("x").rel(("role", "y")).rel("z")` is translated into `RelationConstraint::from(x).add(y).add(z)`. `x`, `y`, and `z` are `RolePlayerConstraint`s at that point, constructed from arguments to `rel()`